nixos/modules/services/backup/bupstash.nix

```
@@ -0,0 +1,501 @@
+                            The send log serves two main purposes:
+
+                            -  it remembers the ids of data chunks that were sent to the repository in the last 'put', allowing bupstash to avoid resending those chunks over the network repeatedly.
+                            -  It stores a mapping of file paths to data that has already been sent, allowing bupstash to skip processing files when snapshotting the same directory many times repeatedly.
+                          '';
+                        };
+                      };
+
+                      config = {
+                        command = mkIf (config.kind == "ssh") (
+                          (s: "ssh -i ${s.privateKey} ${s.user}@${s.server} bupstash serve ${s.path}") config.ssh
+                        );
+                      };
+                    }
+                  )
+                );
+                default = [ ];
+                description = "List of repositories to backup to.";
+              };
+
+              tags = mkOption {
+                type = attrsOf str;
+                default = {
+                  inherit name;
+                };
+                description = ''
+                  Key/value pairs that are applied as tags for this job.
+                  Valid tag keys must match the regular expression ^([a-zA-Z0-9\\-_]+)=(.+)$,
+                  that means tag keys must be alpha numeric with the addition of - and _
+                '';
+              };
+
+              startAt = mkOption {
+                type = either str (listOf str);
+                default = [ ];
+                description = ''
+                  Automatically start this backup at the given date/time, which
+                  must be in the format described in {manpage}`systemd.time(7)`.
+                '';
+                example = "*-*-* 02:30:00";
+              };
+
+              extraArgs = mkOption {
+                type = listOf str;
+                default = [ ];
+                description = ''
+                  Additional options passed to the `bupstash put` command line.
+                  Check the [documentation](https://bupstash.io/doc/man/bupstash-put.html) for possible options.
+                '';
+              };
+            };
+          }
+        )
+      );
+      default = { };
+      description = "Bupstash backup job configurations.";
+    };
+  };
+
+  config = mkMerge [
+    (mkIf (cfg.jobs != { }) {
+      assertions =
+        mapAttrsToList
+          (name: job: {
+            assertion = (job.key == null) != (job.keyCommand == null);
+            message = "Exactly one of key and keyCommand must be provided for the ${name} job.";
+          })
+          cfg.jobs;
+
+      systemd = {
+        services = builtins.listToAttrs (builtins.concatLists (mapAttrsToList mkJobUnits cfg.jobs));
```

I'd like to use `concatMapAttrs` here. However, this would require changing the function signature of `mkJobUnits` to return an attrset instead of a list of namevaluepairs. As a start, you could wrap the current function in `listToAttrs` and then inline that into the function itself later on. (While you could just wrap the entire `mkJobUnits` body in `listToAttrs`, it would be nicer to have it output attribute sets directly. For the first three static items this should be easy, for the list merging you'd have to find appropriate lib functions)

```suggestion
        services = concatMapAttrs (n: v: lib.listToAttrs (mkJobUnits n v)) cfg.jobs;
```

Also, please prefer using the lib functions over the builtins.